### ISSUE TITLE
Unexpected behavior when a function which involves tf.reshape is run using strategy.run on a TPU

### ISSUE THREAD

**Issue by aurelio-amerio** [tensorflow/tensorflow#45851]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- TensorFlow version (use command below): 2.4.0

**Describe the current behavior**
When computing a function which involves `tf.reshape` using `strategy.run` and a TPU, there is an unexpected behavior and errors are thrown. The function works fine if called directly.
This is the same error I get when I use that function as a loss function in a model.

**Describe the expected behavior**
I would expect the same behavior when the function is called directly and using strategy.run. Is there something I am missing out?

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/18yHVyEwbbfYXJ_r3XfP_NaStiVHqwkzN?usp=sharing

I am not completely sure this is a bug of Tensorflow and not a bug in my code, if you could help me I would really appreciate it.
Thank you very much for your help

**Comment by amahendrakar** [comment]
@aurelio-amerio,
Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/c0f92ce5142c9a12b7ec9d883132bed9/45851-2-3.ipynb), [TF v2.4](https://colab.research.google.com/gist/amahendrakar/0b157e29d05cd9b67438827b5bc435d6/45851.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/2b099f98b5b9a95e00a978d32ba807f0/45851-2-5.ipynb). Please check the linked gist for reference. Thanks!

**Comment by aurelio-amerio** [comment]
I have solved the problem on tf 2.4 using `tf.map_fn`. It should work also on the other versions, though I have not tested it. Here is the working notebook for future reference:
https://colab.research.google.com/gist/aurelio-amerio/a15e8646fbb7877120a2f9f33176fd95/test_amerio_26_12_2020.ipynb
